File: cli/fosscut/src/main/java/com/fosscut/type/Order.java
package com.fosscut.type;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class Order {
    private List<Input> inputs;
    private List<Output> outputs;

    public List<Input> getInputs() {
        return this.inputs;
    }

    public void setInputs(List<Input> inputs) {
        this.inputs = inputs;
    }

    public List<Output> getOutputs() {
        return this.outputs;
    }

    public void setOutputs(List<Output> outputs) {
        this.outputs = outputs;
    }

    public boolean isValid() {
        Input longestInput = Collections.max(this.inputs, Comparator.comparing(i -> i.getLength()));
        Output longestOutput = Collections.max(this.outputs, Comparator.comparing(i -> i.getLength()));
        return longestInput.getLength() >= longestOutput.getLength();
    }

    public void validate() {
        if (!this.isValid()) {
            System.err.println("Order invalid. Longest input element must be longer than longest output element.");
            System.exit(1);
        }
    }
}

File: gui/client/src/type/Order.ts
import Input from "./Input.ts";
import Output from "./Output.ts";

class Order {
  inputs: Input[] = [];
  outputs: Output[] = [];

  constructor(order: { inputs?: Input[], outputs?: Output[] }) {
    if (order.inputs && order.outputs) {
      this.inputs = this.generateInputIds(order.inputs) as Input[];
      this.outputs = this.generateOutputIds(order.outputs) as Output[];
    } else {
      if (!order.inputs && !order.outputs) {
        throw new SyntaxError("Incorrect format of order elements. Order must contain both inputs and outputs.");
      } else if (!order.inputs) {
        throw new SyntaxError("Incorrect format of input elements.");
      } else if (!order.outputs) {
        throw new SyntaxError("Incorrect format of output elements.");
      }
    }
  }

  generateInputIds(elements: Input[]): Input[] {
    return elements.map((element, i) => ({ ...element, id: i }) as Input);
  }

  generateOutputIds(elements: Output[]): Output[] {
    return elements.map((element, i) => ({ ...element, id: i }) as Output);
  }

  getInputs(): Input[] {
    return this.inputs;
  }

  setInputs(inputs: Input[]): void {
    this.inputs = inputs;
  }

  getOutputs(): Output[] {
    return this.outputs;
  }

  setOutputs(outputs: Output[]): void {
    this.outputs = outputs;
  }

}

export default Order;
